[Bug]: cannot compile

### Summary

2 compile error: could not compile `zeroclaw` (lib) due to 2 previous errors; 16 warnings emitted

### Affected component

runtime/daemon

### Severity

S1 - workflow blocked

### Current behavior

compile error

### Expected behavior

compile success

### Steps to reproduce

```bash
cargo install --path . --force --locked
```

### Impact

cannot compile

### Logs / stack traces

```text
cargo build --release --all-features
   Compiling zeroclaw v0.1.7 (/data00/home/xieyuanhang/zeroclaw)
error[E0428]: the name `default_otp_enabled` is defined multiple times
    --> src/config/schema.rs:4561:1
     |
4557 | fn default_otp_enabled() -> bool {
     | -------------------------------- previous definition of the value `default_otp_enabled` here
...
4561 | fn default_otp_enabled() -> bool {
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `default_otp_enabled` redefined here
     |
     = note: `default_otp_enabled` must be defined only once in the value namespace of this module

error[E0433]: failed to resolve: use of unresolved module or unlinked crate `tokio_postgres_rustls`
   --> src/memory/postgres.rs:139:31
    |
139 |                     let tls = tokio_postgres_rustls::MakeRustlsConnect::new(tls_config);
    |                               ^^^^^^^^^^^^^^^^^^^^^ use of unresolved module or unlinked crate `tokio_postgres_rustls`
    |
    = help: if you wanted to use a crate named `tokio_postgres_rustls`, use `cargo add tokio_postgres_rustls` to add it to your `Cargo.toml`

warning: unused import: `run_tool_call_loop`
  --> src/gateway/ws.rs:14:74
   |
14 |     build_shell_policy_instructions, build_tool_instructions_from_specs, run_tool_call_loop,
   |                                                                          ^^^^^^^^^^^^^^^^^^
   |
   = note: `#[warn(unused_imports)]` (part of `#[warn(unused)]`) on by default

warning: unexpected `cfg` condition value: `firecrawl`
    --> src/onboard/wizard.rs:3056:19
     |
3056 |             #[cfg(feature = "firecrawl")]
     |                   ^^^^^^^^^^^^^^^^^^^^^
     |
     = note: expected values for `feature` are: `browser-native`, `channel-lark`, `channel-matrix`, `default`, `fantoccini`, `hardware`, `landlock`, `memory-postgres`, `nusb`, `observability-otel`, `peripheral-rpi`, `probe`, `rag-pdf`, `rppal`, `runtime-wasm`, `sandbox-bubblewrap`, `sandbox-landlock`, `tokio-serial`, `wasm-tools`, and `whatsapp-web`
     = help: consider adding `firecrawl` as a feature in `Cargo.toml`
     = note: see <https://doc.rust-lang.org/nightly/rustc/check-cfg/cargo-specifics.html> for more information about checking conditional configuration
     = note: `#[warn(unexpected_cfgs)]` on by default

warning: unexpected `cfg` condition value: `firecrawl`
    --> src/onboard/wizard.rs:3075:19
     |
3075 |             #[cfg(feature = "firecrawl")]
     |                   ^^^^^^^^^^^^^^^^^^^^^
     |
     = note: expected values for `feature` are: `browser-native`, `channel-lark`, `channel-matrix`, `default`, `fantoccini`, `hardware`, `landlock`, `memory-postgres`, `nusb`, `observability-otel`, `peripheral-rpi`, `probe`, `rag-pdf`, `rppal`, `runtime-wasm`, `sandbox-bubblewrap`, `sandbox-landlock`, `tokio-serial`, `wasm-tools`, and `whatsapp-web`
     = help: consider adding `firecrawl` as a feature in `Cargo.toml`
     = note: see <https://doc.rust-lang.org/nightly/rustc/check-cfg/cargo-specifics.html> for more information about checking conditional configuration

warning: unexpected `cfg` condition value: `firecrawl`
    --> src/onboard/wizard.rs:3127:19
     |
3127 |             #[cfg(feature = "firecrawl")]
     |                   ^^^^^^^^^^^^^^^^^^^^^
     |
     = note: expected values for `feature` are: `browser-native`, `channel-lark`, `channel-matrix`, `default`, `fantoccini`, `hardware`, `landlock`, `memory-postgres`, `nusb`, `observability-otel`, `peripheral-rpi`, `probe`, `rag-pdf`, `rppal`, `runtime-wasm`, `sandbox-bubblewrap`, `sandbox-landlock`, `tokio-serial`, `wasm-tools`, and `whatsapp-web`
     = help: consider adding `firecrawl` as a feature in `Cargo.toml`
     = note: see <https://doc.rust-lang.org/nightly/rustc/check-cfg/cargo-specifics.html> for more information about checking conditional configuration

warning: unexpected `cfg` condition value: `firecrawl`
    --> src/onboard/wizard.rs:3140:19
     |
3140 |             #[cfg(feature = "firecrawl")]
     |                   ^^^^^^^^^^^^^^^^^^^^^
     |
     = note: expected values for `feature` are: `browser-native`, `channel-lark`, `channel-matrix`, `default`, `fantoccini`, `hardware`, `landlock`, `memory-postgres`, `nusb`, `observability-otel`, `peripheral-rpi`, `probe`, `rag-pdf`, `rppal`, `runtime-wasm`, `sandbox-bubblewrap`, `sandbox-landlock`, `tokio-serial`, `wasm-tools`, and `whatsapp-web`
     = help: consider adding `firecrawl` as a feature in `Cargo.toml`
     = note: see <https://doc.rust-lang.org/nightly/rustc/check-cfg/cargo-specifics.html> for more information about checking conditional configuration

warning: unused import: `zip::ZipArchive`
 --> src/skills/audit.rs:6:5
  |
6 | use zip::ZipArchive;
  |     ^^^^^^^^^^^^^^^

warning: unexpected `cfg` condition value: `firecrawl`
   --> src/tools/web_fetch.rs:202:11
    |
202 |     #[cfg(feature = "firecrawl")]
    |           ^^^^^^^^^^^^^^^^^^^^^
    |
    = note: expected values for `feature` are: `browser-native`, `channel-lark`, `channel-matrix`, `default`, `fantoccini`, `hardware`, `landlock`, `memory-postgres`, `nusb`, `observability-otel`, `peripheral-rpi`, `probe`, `rag-pdf`, `rppal`, `runtime-wasm`, `sandbox-bubblewrap`, `sandbox-landlock`, `tokio-serial`, `wasm-tools`, and `whatsapp-web`
    = help: consider adding `firecrawl` as a feature in `Cargo.toml`
    = note: see <https://doc.rust-lang.org/nightly/rustc/check-cfg/cargo-specifics.html> for more information about checking conditional configuration

warning: unexpected `cfg` condition value: `firecrawl`
   --> src/tools/web_fetch.rs:279:15
    |
279 |     #[cfg(not(feature = "firecrawl"))]
    |               ^^^^^^^^^^^^^^^^^^^^^
    |
    = note: expected values for `feature` are: `browser-native`, `channel-lark`, `channel-matrix`, `default`, `fantoccini`, `hardware`, `landlock`, `memory-postgres`, `nusb`, `observability-otel`, `peripheral-rpi`, `probe`, `rag-pdf`, `rppal`, `runtime-wasm`, `sandbox-bubblewrap`, `sandbox-landlock`, `tokio-serial`, `wasm-tools`, and `whatsapp-web`
    = help: consider adding `firecrawl` as a feature in `Cargo.toml`
    = note: see <https://doc.rust-lang.org/nightly/rustc/check-cfg/cargo-specifics.html> for more information about checking conditional configuration

warning: unexpected `cfg` condition value: `web-fetch-html2md`
   --> src/tools/web_fetch.rs:106:23
    |
106 |                 #[cfg(feature = "web-fetch-html2md")]
    |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: expected values for `feature` are: `browser-native`, `channel-lark`, `channel-matrix`, `default`, `fantoccini`, `hardware`, `landlock`, `memory-postgres`, `nusb`, `observability-otel`, `peripheral-rpi`, `probe`, `rag-pdf`, `rppal`, `runtime-wasm`, `sandbox-bubblewrap`, `sandbox-landlock`, `tokio-serial`, `wasm-tools`, and `whatsapp-web`
    = help: consider adding `web-fetch-html2md` as a feature in `Cargo.toml`
    = note: see <https://doc.rust-lang.org/nightly/rustc/check-cfg/cargo-specifics.html> for more information about checking conditional configuration

warning: unexpected `cfg` condition value: `web-fetch-html2md`
   --> src/tools/web_fetch.rs:110:27
    |
110 |                 #[cfg(not(feature = "web-fetch-html2md"))]
    |                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: expected values for `feature` are: `browser-native`, `channel-lark`, `channel-matrix`, `default`, `fantoccini`, `hardware`, `landlock`, `memory-postgres`, `nusb`, `observability-otel`, `peripheral-rpi`, `probe`, `rag-pdf`, `rppal`, `runtime-wasm`, `sandbox-bubblewrap`, `sandbox-landlock`, `tokio-serial`, `wasm-tools`, and `whatsapp-web`
    = help: consider adding `web-fetch-html2md` as a feature in `Cargo.toml`
    = note: see <https://doc.rust-lang.org/nightly/rustc/check-cfg/cargo-specifics.html> for more information about checking conditional configuration

warning: unexpected `cfg` condition value: `web-fetch-plaintext`
   --> src/tools/web_fetch.rs:118:23
    |
118 |                 #[cfg(feature = "web-fetch-plaintext")]
    |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: expected values for `feature` are: `browser-native`, `channel-lark`, `channel-matrix`, `default`, `fantoccini`, `hardware`, `landlock`, `memory-postgres`, `nusb`, `observability-otel`, `peripheral-rpi`, `probe`, `rag-pdf`, `rppal`, `runtime-wasm`, `sandbox-bubblewrap`, `sandbox-landlock`, `tokio-serial`, `wasm-tools`, and `whatsapp-web`
    = help: consider adding `web-fetch-plaintext` as a feature in `Cargo.toml`
    = note: see <https://doc.rust-lang.org/nightly/rustc/check-cfg/cargo-specifics.html> for more information about checking conditional configuration

warning: unexpected `cfg` condition value: `web-fetch-plaintext`
   --> src/tools/web_fetch.rs:122:27
    |
122 |                 #[cfg(not(feature = "web-fetch-plaintext"))]
    |                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: expected values for `feature` are: `browser-native`, `channel-lark`, `channel-matrix`, `default`, `fantoccini`, `hardware`, `landlock`, `memory-postgres`, `nusb`, `observability-otel`, `peripheral-rpi`, `probe`, `rag-pdf`, `rppal`, `runtime-wasm`, `sandbox-bubblewrap`, `sandbox-landlock`, `tokio-serial`, `wasm-tools`, and `whatsapp-web`
    = help: consider adding `web-fetch-plaintext` as a feature in `Cargo.toml`
    = note: see <https://doc.rust-lang.org/nightly/rustc/check-cfg/cargo-specifics.html> for more information about checking conditional configuration

warning: unexpected `cfg` condition value: `firecrawl`
   --> src/tools/web_search_tool.rs:197:11
    |
197 |     #[cfg(feature = "firecrawl")]
    |           ^^^^^^^^^^^^^^^^^^^^^
    |
    = note: expected values for `feature` are: `browser-native`, `channel-lark`, `channel-matrix`, `default`, `fantoccini`, `hardware`, `landlock`, `memory-postgres`, `nusb`, `observability-otel`, `peripheral-rpi`, `probe`, `rag-pdf`, `rppal`, `runtime-wasm`, `sandbox-bubblewrap`, `sandbox-landlock`, `tokio-serial`, `wasm-tools`, and `whatsapp-web`
    = help: consider adding `firecrawl` as a feature in `Cargo.toml`
    = note: see <https://doc.rust-lang.org/nightly/rustc/check-cfg/cargo-specifics.html> for more information about checking conditional configuration

warning: unexpected `cfg` condition value: `firecrawl`
   --> src/tools/web_search_tool.rs:294:15
    |
294 |     #[cfg(not(feature = "firecrawl"))]
    |               ^^^^^^^^^^^^^^^^^^^^^
    |
    = note: expected values for `feature` are: `browser-native`, `channel-lark`, `channel-matrix`, `default`, `fantoccini`, `hardware`, `landlock`, `memory-postgres`, `nusb`, `observability-otel`, `peripheral-rpi`, `probe`, `rag-pdf`, `rppal`, `runtime-wasm`, `sandbox-bubblewrap`, `sandbox-landlock`, `tokio-serial`, `wasm-tools`, and `whatsapp-web`
    = help: consider adding `firecrawl` as a feature in `Cargo.toml`
    = note: see <https://doc.rust-lang.org/nightly/rustc/check-cfg/cargo-specifics.html> for more information about checking conditional configuration

warning: value assigned to `is_direct_room` is never read
   --> src/channels/matrix.rs:786:42
    |
786 |                 let mut is_direct_room = false;
    |                                          ^^^^^
    |
    = help: maybe it is overwritten before being read?
    = note: `#[warn(unused_assignments)]` (part of `#[warn(unused)]`) on by default

warning: unused variable: `approval_manager`
   --> src/gateway/ws.rs:195:9
    |
195 |     let approval_manager = {
    |         ^^^^^^^^^^^^^^^^ help: if this is intentional, prefix it with an underscore: `_approval_manager`
    |
    = note: `#[warn(unused_variables)]` (part of `#[warn(unused)]`) on by default

Some errors have detailed explanations: E0428, E0433.
For more information about an error, try `rustc --explain E0428`.
warning: `zeroclaw` (lib) generated 16 warnings
error: could not compile `zeroclaw` (lib) due to 2 previous errors; 16 warnings emitted
```

### ZeroClaw version

c0e78ba2bd3d7d1c23dc3e3009c2ed2289b32d4c

### Rust version

rustc 1.93.1

### Operating system

Debian 12

### Regression?

Yes, it worked before

### Pre-flight checks

- [x] I reproduced this on the latest main branch or latest release.
- [x] I redacted secrets/tokens from logs.
- [x] I removed personal identifiers and replaced identity-specific data with neutral placeholders.

--- a/src/config/schema.rs
+++ b/src/config/schema.rs
@@ -1429,10 +1429,11 @@ pub struct WebFetchConfig {
     /// Enable `web_fetch` tool for fetching web page content
     #[serde(default)]
     pub enabled: bool,
-    /// Provider: "fast_html2md", "nanohtml2text", or "firecrawl"
+    /// Provider: "fast_html2md", "nanohtml2text", "firecrawl", or "tavily"
     #[serde(default = "default_web_fetch_provider")]
     pub provider: String,
-    /// Optional provider API key (required for provider = "firecrawl")
+    /// Optional provider API key (required for provider = "firecrawl" or "tavily").
+    /// Multiple keys can be comma-separated for round-robin load balancing.
     #[serde(default)]
     pub api_key: Option<String>,
     /// Optional provider API URL override (for self-hosted providers)
@@ -1491,10 +1492,11 @@ pub struct WebSearchConfig {
     /// Enable `web_search_tool` for web searches
     #[serde(default)]
     pub enabled: bool,
-    /// Search provider: "duckduckgo" (free, no API key) or "brave" (requires API key)
+    /// Search provider: "duckduckgo" (free, no API key), "brave", "firecrawl", or "tavily"
     #[serde(default = "default_web_search_provider")]
     pub provider: String,
-    /// Generic provider API key (used by firecrawl and as fallback for brave)
+    /// Generic provider API key (used by firecrawl, tavily, and as fallback for brave).
+    /// Multiple keys can be comma-separated for round-robin load balancing.
     #[serde(default)]
     pub api_key: Option<String>,
     /// Optional provider API URL override (for self-hosted providers)

--- a/src/tools/web_fetch.rs
+++ b/src/tools/web_fetch.rs
@@ -6,6 +6,7 @@ use crate::config::UrlAccessConfig;
 use crate::security::SecurityPolicy;
 use async_trait::async_trait;
 use serde_json::json;
+use std::sync::atomic::{AtomicUsize, Ordering};
 use std::sync::Arc;
 use std::time::Duration;
 
@@ -15,10 +16,11 @@ use std::time::Duration;
 /// - `fast_html2md`: fetch with reqwest, convert HTML to markdown
 /// - `nanohtml2text`: fetch with reqwest, convert HTML to plaintext
 /// - `firecrawl`: fetch using Firecrawl cloud/self-hosted API
+/// - `tavily`: fetch using Tavily Extract API
 pub struct WebFetchTool {
     security: Arc<SecurityPolicy>,
     provider: String,
-    api_key: Option<String>,
+    api_keys: Vec<String>,
     api_url: Option<String>,
     allowed_domains: Vec<String>,
     blocked_domains: Vec<String>,
@@ -26,6 +28,7 @@ pub struct WebFetchTool {
     max_response_size: usize,
     timeout_secs: u64,
     user_agent: String,
+    key_index: Arc<AtomicUsize>,
 }
 
 impl WebFetchTool {
@@ -43,6 +46,16 @@ impl WebFetchTool {
         user_agent: String,
     ) -> Self {
         let provider = provider.trim().to_lowercase();
+        let api_keys = api_key
+            .as_ref()
+            .map(|raw| {
+                raw.split(',')
+                    .map(str::trim)
+                    .filter(|s| !s.is_empty())
+                    .map(ToOwned::to_owned)
+                    .collect()
+            })
+            .unwrap_or_default();
         Self {
             security,
             provider: if provider.is_empty() {
@@ -50,7 +63,7 @@ impl WebFetchTool {
             } else {
                 provider
             },
-            api_key,
+            api_keys,
             api_url,
             allowed_domains: normalize_allowed_domains(allowed_domains),
             blocked_domains: normalize_allowed_domains(blocked_domains),
@@ -58,7 +71,16 @@ impl WebFetchTool {
             max_response_size,
             timeout_secs,
             user_agent,
+            key_index: Arc::new(AtomicUsize::new(0)),
         }
+    }
+
+    fn get_next_api_key(&self) -> Option<String> {
+        if self.api_keys.is_empty() {
+            return None;
+        }
+        let idx = self.key_index.fetch_add(1, Ordering::Relaxed) % self.api_keys.len();
+        Some(self.api_keys[idx].clone())
     }
 
     fn validate_url(&self, raw_url: &str) -> anyhow::Result<String> {
@@ -127,7 +149,7 @@ impl WebFetchTool {
                 }
             }
             _ => anyhow::bail!(
-                "Unknown web_fetch provider: '{}'. Set tools.web_fetch.provider to 'fast_html2md', 'nanohtml2text', or 'firecrawl' in config.toml",
+                "Unknown web_fetch provider: '{}'. Set [web_fetch].provider to 'fast_html2md', 'nanohtml2text', 'firecrawl', or 'tavily' in config.toml",
                 self.provider
             ),
         }
@@ -201,14 +223,11 @@ impl WebFetchTool {
 
     #[cfg(feature = "firecrawl")]
     async fn fetch_with_firecrawl(&self, url: &str) -> anyhow::Result<String> {
-        let auth_token = match self.api_key.as_ref() {
-            Some(raw) if !raw.trim().is_empty() => raw.trim(),
-            _ => {
-                anyhow::bail!(
-                    "web_fetch provider 'firecrawl' requires [web_fetch].api_key in config.toml"
-                );
-            }
-        };
+        let auth_token = self.get_next_api_key().ok_or_else(|| {
+            anyhow::anyhow!(
+                "web_fetch provider 'firecrawl' requires [web_fetch].api_key in config.toml"
+            )
+        })?;
 
         let api_url = self
             .api_url
@@ -223,7 +242,7 @@ impl WebFetchTool {
             .post(endpoint)
             .header(
                 reqwest::header::AUTHORIZATION,
-                format!("Bearer {auth_token}"),
+                format!("Bearer {}", auth_token),
             )
             .json(&json!({
                 "url": url,
@@ -281,6 +300,71 @@ impl WebFetchTool {
     async fn fetch_with_firecrawl(&self, _url: &str) -> anyhow::Result<String> {
         anyhow::bail!("web_fetch provider 'firecrawl' requires Cargo feature 'firecrawl'")
     }
+
+    async fn fetch_with_tavily(&self, url: &str) -> anyhow::Result<String> {
+        let api_key = self.get_next_api_key().ok_or_else(|| {
+            anyhow::anyhow!(
+                "web_fetch provider 'tavily' requires [web_fetch].api_key in config.toml"
+            )
+        })?;
+
+        let api_url = self
+            .api_url
+            .as_deref()
+            .map(str::trim)
+            .filter(|s| !s.is_empty())
+            .unwrap_or("https://api.tavily.com");
+        let endpoint = format!("{}/extract", api_url.trim_end_matches('/'));
+
+        let response = self
+            .build_http_client()?
+            .post(endpoint)
+            .json(&json!({
+                "api_key": api_key,
+                "urls": [url]
+            }))
+            .send()
+            .await?;
+        let status = response.status();
+        let body = response.text().await?;
+
+        if !status.is_success() {
+            anyhow::bail!(
+                "Tavily extract failed with status {}: {}",
+                status.as_u16(),
+                body
+            );
+        }
+
+        let parsed: serde_json::Value = serde_json::from_str(&body)
+            .map_err(|e| anyhow::anyhow!("Invalid Tavily response JSON: {e}"))?;
+        if let Some(error) = parsed.get("error").and_then(serde_json::Value::as_str) {
+            anyhow::bail!("Tavily API error: {error}");
+        }
+
+        let results = parsed
+            .get("results")
+            .and_then(serde_json::Value::as_array)
+            .ok_or_else(|| anyhow::anyhow!("Tavily response missing results array"))?;
+        if results.is_empty() {
+            anyhow::bail!("Tavily returned no results for URL: {}", url);
+        }
+
+        let result = &results[0];
+        let output = result
+            .get("raw_content")
+            .and_then(serde_json::Value::as_str)
+            .or_else(|| result.get("content").and_then(serde_json::Value::as_str))
+            .unwrap_or("")
+            .trim()
+            .to_string();
+
+        if output.is_empty() {
+            anyhow::bail!("Tavily returned empty content for URL: {}", url);
+        }
+
+        Ok(output)
+    }
 }
 
 #[async_trait]
@@ -290,7 +374,7 @@ impl Tool for WebFetchTool {
     }
 
     fn description(&self) -> &str {
-        "Fetch a web page and return markdown/text content for LLM consumption. Providers: fast_html2md, nanohtml2text, firecrawl. Security: allowlist-only domains, blocked_domains, and no local/private hosts."
+        "Fetch a web page and return markdown/text content for LLM consumption. Providers: fast_html2md, nanohtml2text, firecrawl, tavily. Security: allowlist-only domains, blocked_domains, and no local/private hosts."
     }
 
     fn parameters_schema(&self) -> serde_json::Value {
@@ -342,8 +426,9 @@ impl Tool for WebFetchTool {
         let result = match self.provider.as_str() {
             "fast_html2md" | "nanohtml2text" => self.fetch_with_http_provider(&url).await,
             "firecrawl" => self.fetch_with_firecrawl(&url).await,
+            "tavily" => self.fetch_with_tavily(&url).await,
             _ => Err(anyhow::anyhow!(
-                "Unknown web_fetch provider: '{}'. Set tools.web_fetch.provider to 'fast_html2md', 'nanohtml2text', or 'firecrawl' in config.toml",
+                "Unknown web_fetch provider: '{}'. Set [web_fetch].provider to 'fast_html2md', 'nanohtml2text', 'firecrawl', or 'tavily' in config.toml",
                 self.provider
             )),
         };
@@ -705,5 +790,32 @@ mod tests {
         } else {
             assert!(error.contains("requires Cargo feature 'firecrawl'"));
         }
+    }
+
+    #[tokio::test]
+    async fn tavily_provider_requires_api_key() {
+        let tool = test_tool_with_provider(vec!["*"], vec![], "tavily", None, None);
+        let result = tool
+            .execute(json!({"url": "https://example.com"}))
+            .await
+            .unwrap();
+        assert!(!result.success);
+        let error = result.error.unwrap_or_default();
+        assert!(error.contains("requires [web_fetch].api_key"));
+    }
+
+    #[test]
+    fn parses_multiple_api_keys() {
+        let tool =
+            test_tool_with_provider(vec!["*"], vec![], "tavily", Some("key1,key2,key3"), None);
+        assert_eq!(tool.api_keys, vec!["key1", "key2", "key3"]);
+    }
+
+    #[test]
+    fn round_robin_api_key_selection_cycles() {
+        let tool = test_tool_with_provider(vec!["*"], vec![], "tavily", Some("k1,k2"), None);
+        assert_eq!(tool.get_next_api_key().as_deref(), Some("k1"));
+        assert_eq!(tool.get_next_api_key().as_deref(), Some("k2"));
+        assert_eq!(tool.get_next_api_key().as_deref(), Some("k1"));
     }
 }

--- a/src/tools/web_search_tool.rs
+++ b/src/tools/web_search_tool.rs
@@ -4,19 +4,21 @@ use async_trait::async_trait;
 use regex::Regex;
 use reqwest::StatusCode;
 use serde_json::json;
+use std::sync::atomic::{AtomicUsize, Ordering};
 use std::sync::Arc;
 use std::time::Duration;
 
 /// Web search tool for searching the internet.
-/// Supports providers: DuckDuckGo (free), Brave, Firecrawl.
+/// Supports providers: DuckDuckGo (free), Brave, Firecrawl, Tavily.
 pub struct WebSearchTool {
     security: Arc<SecurityPolicy>,
     provider: String,
-    api_key: Option<String>,
+    api_keys: Vec<String>,
     api_url: Option<String>,
     max_results: usize,
     timeout_secs: u64,
     user_agent: String,
+    key_index: Arc<AtomicUsize>,
 }
 
 impl WebSearchTool {
@@ -41,15 +43,34 @@ impl WebSearchTool {
         timeout_secs: u64,
         user_agent: String,
     ) -> Self {
+        let api_keys = api_key
+            .as_ref()
+            .map(|raw| {
+                raw.split(',')
+                    .map(str::trim)
+                    .filter(|s| !s.is_empty())
+                    .map(ToOwned::to_owned)
+                    .collect()
+            })
+            .unwrap_or_default();
         Self {
             security,
             provider: provider.trim().to_lowercase(),
-            api_key,
+            api_keys,
             api_url,
             max_results: max_results.clamp(1, 10),
             timeout_secs: timeout_secs.max(1),
             user_agent,
+            key_index: Arc::new(AtomicUsize::new(0)),
         }
+    }
+
+    fn get_next_api_key(&self) -> Option<String> {
+        if self.api_keys.is_empty() {
+            return None;
+        }
+        let idx = self.key_index.fetch_add(1, Ordering::Relaxed) % self.api_keys.len();
+        Some(self.api_keys[idx].clone())
     }
 
     async fn search_duckduckgo(&self, query: &str) -> anyhow::Result<String> {
@@ -129,10 +150,9 @@ impl WebSearchTool {
     }
 
     async fn search_brave(&self, query: &str) -> anyhow::Result<String> {
-        let auth_token = match self.api_key.as_ref() {
-            Some(raw) if !raw.trim().is_empty() => raw.trim(),
-            _ => anyhow::bail!("Brave API key not configured"),
-        };
+        let auth_token = self
+            .get_next_api_key()
+            .ok_or_else(|| anyhow::anyhow!("Brave API key not configured"))?;
 
         let encoded_query = urlencoding::encode(query);
         let search_url = format!(
@@ -196,14 +216,11 @@ impl WebSearchTool {
 
     #[cfg(feature = "firecrawl")]
     async fn search_firecrawl(&self, query: &str) -> anyhow::Result<String> {
-        let auth_token = match self.api_key.as_ref() {
-            Some(raw) if !raw.trim().is_empty() => raw.trim(),
-            _ => {
-                anyhow::bail!(
-                    "web_search provider 'firecrawl' requires [web_search].api_key in config.toml"
-                );
-            }
-        };
+        let auth_token = self.get_next_api_key().ok_or_else(|| {
+            anyhow::anyhow!(
+                "web_search provider 'firecrawl' requires [web_search].api_key in config.toml"
+            )
+        })?;
 
         let api_url = self
             .api_url
@@ -221,7 +238,7 @@ impl WebSearchTool {
             .post(endpoint)
             .header(
                 reqwest::header::AUTHORIZATION,
-                format!("Bearer {auth_token}"),
+                format!("Bearer {}", auth_token),
             )
             .json(&json!({
                 "query": query,
@@ -295,6 +312,89 @@ impl WebSearchTool {
     #[allow(clippy::unused_async)]
     async fn search_firecrawl(&self, _query: &str) -> anyhow::Result<String> {
         anyhow::bail!("web_search provider 'firecrawl' requires Cargo feature 'firecrawl'")
+    }
+
+    async fn search_tavily(&self, query: &str) -> anyhow::Result<String> {
+        let api_key = self.get_next_api_key().ok_or_else(|| {
+            anyhow::anyhow!(
+                "web_search provider 'tavily' requires [web_search].api_key in config.toml"
+            )
+        })?;
+
+        let api_url = self
+            .api_url
+            .as_deref()
+            .map(str::trim)
+            .filter(|s| !s.is_empty())
+            .unwrap_or("https://api.tavily.com");
+        let endpoint = format!("{}/search", api_url.trim_end_matches('/'));
+
+        let client = reqwest::Client::builder()
+            .timeout(Duration::from_secs(self.timeout_secs))
+            .user_agent(self.user_agent.as_str())
+            .build()?;
+        let response = client
+            .post(&endpoint)
+            .json(&json!({
+                "api_key": api_key,
+                "query": query,
+                "max_results": self.max_results,
+                "search_depth": "basic",
+                "include_answer": false,
+                "include_raw_content": false,
+                "include_images": false
+            }))
+            .send()
+            .await
+            .map_err(|e| anyhow::anyhow!("Tavily search failed: {e}"))?;
+        let status = response.status();
+        let body = response.text().await?;
+        if !status.is_success() {
+            anyhow::bail!(
+                "Tavily search failed with status {}: {}",
+                status.as_u16(),
+                body
+            );
+        }
+
+        let parsed: serde_json::Value = serde_json::from_str(&body)
+            .map_err(|e| anyhow::anyhow!("Invalid Tavily response JSON: {e}"))?;
+        if let Some(error) = parsed.get("error").and_then(serde_json::Value::as_str) {
+            anyhow::bail!("Tavily API error: {error}");
+        }
+
+        let results = parsed
+            .get("results")
+            .and_then(serde_json::Value::as_array)
+            .ok_or_else(|| anyhow::anyhow!("Tavily response missing results array"))?;
+        if results.is_empty() {
+            return Ok(format!("No results found for: {}", query));
+        }
+
+        let mut lines = vec![format!("Search results for: {} (via Tavily)", query)];
+        for (i, result) in results.iter().take(self.max_results).enumerate() {
+            let title = result
+                .get("title")
+                .and_then(serde_json::Value::as_str)
+                .unwrap_or("No title");
+            let url = result
+                .get("url")
+                .and_then(serde_json::Value::as_str)
+                .unwrap_or("");
+            let content = result
+                .get("content")
+                .and_then(serde_json::Value::as_str)
+                .unwrap_or("")
+                .trim();
+
+            lines.push(format!("{}. {}", i + 1, title));
+            lines.push(format!("   {}", url));
+            if !content.is_empty() {
+                lines.push(format!("   {}", content));
+            }
+        }
+
+        Ok(lines.join("\n"))
     }
 }
 
@@ -370,8 +470,9 @@ impl Tool for WebSearchTool {
             "duckduckgo" | "ddg" => self.search_duckduckgo(query).await?,
             "brave" => self.search_brave(query).await?,
             "firecrawl" => self.search_firecrawl(query).await?,
+            "tavily" => self.search_tavily(query).await?,
             _ => anyhow::bail!(
-                "Unknown search provider: '{}'. Set tools.web_search.provider to 'duckduckgo', 'brave', or 'firecrawl' in config.toml",
+                "Unknown search provider: '{}'. Set [web_search].provider to 'duckduckgo', 'brave', 'firecrawl', or 'tavily' in config.toml",
                 self.provider
             ),
         };
@@ -601,6 +702,52 @@ mod tests {
         } else {
             assert!(error.contains("requires Cargo feature 'firecrawl'"));
         }
+    }
+
+    #[tokio::test]
+    async fn test_execute_tavily_without_api_key() {
+        let tool = WebSearchTool::new(
+            test_security(),
+            "tavily".to_string(),
+            None,
+            None,
+            5,
+            15,
+            "test".to_string(),
+        );
+        let result = tool.execute(json!({"query": "test"})).await;
+        assert!(result.is_err());
+        assert!(result.unwrap_err().to_string().contains("api_key"));
+    }
+
+    #[test]
+    fn test_parses_multiple_api_keys() {
+        let tool = WebSearchTool::new(
+            test_security(),
+            "tavily".to_string(),
+            Some("key1,key2,key3".to_string()),
+            None,
+            5,
+            15,
+            "test".to_string(),
+        );
+        assert_eq!(tool.api_keys, vec!["key1", "key2", "key3"]);
+    }
+
+    #[test]
+    fn test_round_robin_api_key_selection_cycles() {
+        let tool = WebSearchTool::new(
+            test_security(),
+            "tavily".to_string(),
+            Some("k1,k2".to_string()),
+            None,
+            5,
+            15,
+            "test".to_string(),
+        );
+        assert_eq!(tool.get_next_api_key().as_deref(), Some("k1"));
+        assert_eq!(tool.get_next_api_key().as_deref(), Some("k2"));
+        assert_eq!(tool.get_next_api_key().as_deref(), Some("k1"));
     }
 
     #[tokio::test]

